File: app/src/main/java/com/mdmobile/cyclops/networkRequests/BasicRequest.java
package com.mdmobile.cyclops.networkRequests;


import android.accounts.Account;
import android.accounts.AccountManager;
import android.os.Bundle;
import android.util.Log;

import com.android.volley.AuthFailureError;
import com.android.volley.NetworkResponse;
import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.HttpHeaderParser;
import com.mdmobile.cyclops.R;
import com.mdmobile.cyclops.interfaces.OnTokenAcquired;
import com.mdmobile.cyclops.provider.McContract;
import com.mdmobile.cyclops.utils.Logger;
import com.mdmobile.cyclops.utils.ServerUtility;
import com.mdmobile.cyclops.utils.UserUtility;

import java.io.UnsupportedEncodingException;
import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.Map;

import javax.net.ssl.HttpsURLConnection;

import static com.mdmobile.cyclops.ApplicationLoader.applicationContext;
import static com.mdmobile.cyclops.services.AccountAuthenticator.AUTH_TOKEN_TYPE_KEY;

/**
 * Extend this class in Custom volley request.
 * This insert MobiControl authentication token in the headers
 * And provides standard procedure for a refused request
 */

abstract public class BasicRequest<T> extends Request<T> {

    private final String LOG_TAG = BasicRequest.class.getSimpleName();

    public BasicRequest(int method, String url, Response.ErrorListener errorListener) {
        super(method, url, errorListener);
        setRetryPolicy(new BasicRequestRetry(this));
    }

    @Override
    public Map<String, String> getHeaders() throws AuthFailureError {
        Map<String, String> headers = new HashMap<>();
        AccountManager accountManager = AccountManager.get(applicationContext);

        //TODO: support multiple account
        Account account = UserUtility.getUser();
        String tokenType = UserUtility.getUserInfo(UserUtility.getUser()).getString(AUTH_TOKEN_TYPE_KEY);

        String token = accountManager.peekAuthToken(account, tokenType);

        if (token == null) {
            //TODO: stop request and launch account manager for a new token
            //Instead of sending this request out which is gonna fail because of the token cancel this request
            //and start the new token request procedure
            return super.getHeaders();
        }

        headers.put("Authorization", tokenType.concat(" ").concat(token));
        headers.put("Accept", "application/json");

        return headers;
    }

    @Override
    protected VolleyError parseNetworkError(VolleyError volleyError) {

        //TODO:support Multiple account

        NetworkResponse response = volleyError.networkResponse;
        if (response == null) {
            Logger.log(LOG_TAG, "Network request Error: error response: NULL\n", Log.ERROR);
            return super.parseNetworkError(volleyError);
        }

        try {
            String errorResponse = new String(response.data, HttpHeaderParser.parseCharset(response.headers));
            Logger.log(LOG_TAG, errorResponse, Log.ERROR);
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }

        //CHECK ALL ERRORS CODE
        parseErrorCode(response.statusCode);
        return super.parseNetworkError(volleyError);
    }

    private void parseErrorCode(int errorCode) {

        if (errorCode == HttpsURLConnection.HTTP_BAD_REQUEST) {
            Logger.log(LOG_TAG, "Error " + errorCode + " HTTP BAD REQUEST", Log.ERROR);
        } else if (errorCode == HttpsURLConnection.HTTP_UNAUTHORIZED ||
                errorCode == HttpsURLConnection.HTTP_FORBIDDEN) {
            Logger.log(LOG_TAG, "Attempt to request new token\n", Log.VERBOSE);

            AccountManager manager = AccountManager.get(applicationContext);
            Account[] accounts = manager.getAccountsByType(applicationContext.getString(R.string.account_type));

            if (accounts.length == 1) {
                String tokenType = manager.getUserData(accounts[0], AUTH_TOKEN_TYPE_KEY);
                String token = manager.peekAuthToken(accounts[0], tokenType);
                manager.invalidateAuthToken(applicationContext.getString(R.string.account_type), token);

                manager.getAuthToken(accounts[0],
                        manager.getUserData(accounts[0], AUTH_TOKEN_TYPE_KEY),
                        null, false, new OnTokenAcquired(new WeakReference<BasicRequest>(this)), null);
            }
        } else if (errorCode == HttpsURLConnection.HTTP_NOT_FOUND) {
            Bundle server = ServerUtility.getServer();
            String serverName;
            if (server != null) {
                serverName = server.getString(McContract.ServerInfo.NAME);
            } else {
                serverName = getUrl();
            }
            ServerUtility.notifyServerStatus(serverName, ServerUtility.SERVER_OFFLINE);
        } else if (errorCode == 422) {
            //TODO: show error message in toast
        } else if (errorCode == HttpsURLConnection.HTTP_INTERNAL_ERROR) {
        }
    }
}

File: app/src/main/java/com/mdmobile/cyclops/networkRequests/BasicRequest.kt
package com.mdmobile.cyclops.networkRequests


import android.accounts.AccountManager
import android.icu.lang.UCharacter.GraphemeClusterBreak.L
import android.net.Uri
import android.util.Log

import com.android.volley.AuthFailureError
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.VolleyError
import com.android.volley.toolbox.HttpHeaderParser
import com.mdmobile.cyclops.security.ServerNotFound
import com.mdmobile.cyclops.util.Logger
import com.mdmobile.cyclops.util.ServerUtility
import com.mdmobile.cyclops.util.UserUtility

import java.io.UnsupportedEncodingException
import java.util.HashMap

import javax.net.ssl.HttpsURLConnection

import com.mdmobile.cyclops.CyclopsApplication.Companion
import com.mdmobile.cyclops.services.AccountAuthenticator
import getMcAccount
import java.nio.charset.Charset

/**
 * Extend this class in Custom volley request.
 * This insert MobiControl authentication _token in the headers
 * And provides standard procedure for a refused request
 */

abstract class BasicRequest<T>(method: Int, url: String, errorListener: Response.ErrorListener)//        setRetryPolicy(new BasicRequestRetry(this));
    : Request<T>(method, url, errorListener) {

     private val  logTag = BasicRequest::class.java.simpleName

    val serverUrl: String?
        get() = Uri.parse(url).authority

    @Throws(AuthFailureError::class)
    override fun getHeaders(): Map<String, String> {
        val headers = HashMap<String, String>()
        val accountManager = AccountManager.get(Companion.applicationContext)

        //TODO: support multiple account
        val account = accountManager.getMcAccount()
        val tokenType = UserUtility.getUserInfo(accountManager.getMcAccount()).getString(AccountAuthenticator.AUTH_TOKEN_TYPE_KEY)

        val token = accountManager.peekAuthToken(account, tokenType)
                ?: //TODO: stop request and launch account manager for a new _token
                //Instead of sending this request out which is gonna fail because of the _token cancel this request
                //and start the new _token request procedure
                return super.getHeaders()

        headers["Authorization"] = "$tokenType $token"
        headers["Accept"] = "application/json"

        return headers
    }

    override fun parseNetworkError(volleyError: VolleyError): VolleyError {

        //TODO:support Multiple account
        val response = volleyError.networkResponse
        if (response == null) {
            Logger.log(logTag, "Network request Error: error response: NULL\n", Log.ERROR)
            return super.parseNetworkError(volleyError)
        }

        try {
            val errorResponse = String(response.data, (HttpHeaderParser.parseCharset(response.headers)).toCharArray() as Charset)
            Logger.log(logTag, errorResponse, Log.ERROR)
            //CHECK ALL ERRORS CODE
            parseErrorCode(response.statusCode)
        } catch (e: UnsupportedEncodingException) {
            e.printStackTrace()
        } catch (e: ServerNotFound) {
            e.printStackTrace()
        }

        return super.parseNetworkError(volleyError)
    }

    @Throws(ServerNotFound::class)
    private fun parseErrorCode(errorCode: Int) {

        if (errorCode == HttpsURLConnection.HTTP_BAD_REQUEST) {
            Logger.log(logTag, "Error $errorCode HTTP BAD REQUEST", Log.ERROR)
        } else if (errorCode == HttpsURLConnection.HTTP_UNAUTHORIZED || errorCode == HttpsURLConnection.HTTP_FORBIDDEN) {
            //            Logger.log(logTag, "Attempt to request new _token\n", Log.VERBOSE);
            //
            //            AccountManager manager = AccountManager.get(applicationContext);
            //            Account[] accounts = manager.getAccountsByType(applicationContext.getString(R.string.account_type));
            //
            //            if (accounts.length == 1) {
            //                String tokenType = manager.getUserData(accounts[0], AUTH_TOKEN_TYPE_KEY);
            //                String _token = manager.peekAuthToken(accounts[0], tokenType);
            //                manager.invalidateAuthToken(applicationContext.getString(R.string.account_type), _token);
            //
            //                manager.getAuthToken(accounts[0],
            //                        manager.getUserData(accounts[0], AUTH_TOKEN_TYPE_KEY), null, true,
            //                        new OnTokenResponse(new WeakReference<BasicRequest>(this)), null);
            //            }

        } else if (errorCode == HttpsURLConnection.HTTP_NOT_FOUND) {

            val serverName = ServerUtility.getActiveServer().serverName
            ServerUtility.notifyServerStatus(serverName, ServerUtility.SERVER_OFFLINE)

        } else if (errorCode == 422) {
            //TODO: show error message in toast
        } else if (errorCode == HttpsURLConnection.HTTP_INTERNAL_ERROR) {
        }
    }
}

